File: 백준/Gold/17298. 오큰수/오큰수.cc
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    int n;
    cin >> n;
    
    vector<int> arr(n);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    vector<int> stack;
    vector<int> ans(n);
    for (int i = n - 1; i >= 0; i--) {
        int h = arr[i];
        while(!stack.empty() && h >= stack.back()) {
            stack.pop_back();
        }
        ans[i] = stack.empty() ? -1 : stack.back();
        stack.push_back(h);
    }
    for (int i = 0; i < n; i++) {
        cout << ans[i] << " ";
    }
    return 0;
}

File: 백준/Gold/17298. 오큰수/오큰수.py
n = int(input())
arr = list(map(int, input().split()))
ans = []
arr = reversed(arr)

stack = []

for num in arr:
    while(stack and num >= stack[-1]):
        stack.pop()
    ans.append(stack[-1] if stack else -1)
    stack.append(num)
print(*reversed(ans))